### PR TITLE
matchers: provide a way to pass config

### PR DESCRIPTION
**Pull request by arajkumar**
This PR implements factory with registry pattern for matchers to facilitate passing configurable parameters to matchers.

### CODE REVIEW

File: libvuln/opts.go [hunk 1]
```diff
@@ -69,10 +60,28 @@ type Opts struct {
...
+	// "suse"
+	// "ubuntu"
+	// "crda" - remotematcher calls hosted api via RPC.
+	MatcherNames []string
+
+	// Config holds configuration blocks for MatcherFactories and Matchers,
+	// keyed by name.
+	MatcherConfigs map[string]driver.MatcherConfigUnmarshaler
+
 	Matchers []driver.Matcher
```
Comment (ldelossa) [inline libvuln/opts.go:85]: This needs comments, its for allowing out of tree updaters to be provided to libvuln.

File: crda/matcherfactory.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"net/url"
+
+	"github.com/quay/zlog"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/label"
+
+	"github.com/quay/claircore/libvuln/driver"
+)
+
+// Factory contains the configuration for fetching and parsing a pulp manifest.
```
Comment (ldelossa) [inline crda/matcherfactory.go:15]: Comments are stale.

File: libvuln/driver/matcherfactory.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// information.
+type MatcherConfigUnmarshaler func(interface{}) error
+
+// MatcherConfigurable is an interface that MatcherFactory can implement to opt-in to having
+// their configuration provided dynamically.
+type MatcherConfigurable interface {
+	Configure(context.Context, MatcherConfigUnmarshaler, *http.Client) error
+}
+
+// MatcherFactoryFunc would ease the registration of Matchers which doesn't
```
Comment (ldelossa) [inline libvuln/driver/matcherfactory.go:27]: s/doesn't/don't

**Review by ldelossa — CHANGES_REQUESTED** [review]

File: matchers/matchers.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	// configs provided to matchers once constructed.
+	configs Configs
+	client  *http.Client
+	// out-of-tree matchers
+	matchers []driver.Matcher
+}
+
+type MatchersOption func(m *Matchers)
+
+// NewManager will return a manager ready to have its Start or Run methods called.
```
Comment (ldelossa) [inline matchers/matchers.go:31]: stale comments. 

Also out of curiosity why did you use the Manager pattern used for a long running update manager here?

File: libvuln/opts.go (hunk 1 above)
Comment (arajkumar) [reply libvuln/opts.go]: Fixed it.

File: crda/matcherfactory.go (hunk 2 above)
Comment (arajkumar) [reply crda/matcherfactory.go]: Fixed it.

File: libvuln/driver/matcherfactory.go (hunk 3 above)
Comment (arajkumar) [reply libvuln/driver/matcherfactory.go]: Fixed it.

File: matchers/matchers.go (hunk 4 above)
Comment (arajkumar) [reply matchers/matchers.go:31]: I've implemented it based on `updates/manager.go`, but it doesn't exhibit the behaviour of manager, it is still a factory of factory.

**Review by ldelossa — APPROVED** [review]

**Pull request merged**